File: backend/app/routers/income.py
"""收入設定（SRS 4.1）。路由只做參數接收與回應，業務規則在 services/income.py。

- GET/PUT /months/{month}/income          月薪與儲蓄目標（GET 無紀錄時沿用上一有紀錄月份，不寫入）
- GET/POST/PUT/DELETE /extra-incomes      額外收入（GET 可帶 ?month=YYYY-MM）
- GET/POST/PUT/DELETE /recurring-expenses 固定支出（GET 可帶 ?month=YYYY-MM 只列該月生效者）

GET /months/{month}（月總覽）屬 Phase 2，不在此。
"""

from fastapi import APIRouter, Response, status

from app.deps import CurrentUserId, DbSession
from app.schemas.common import error_responses
from app.schemas.income import (
    ExtraIncomeCreate,
    ExtraIncomeOut,
    ExtraIncomeUpdate,
    MonthIncomeOut,
    MonthIncomeUpdate,
    RecurringExpenseCreate,
    RecurringExpenseOut,
    RecurringExpenseUpdate,
)
from app.services import income as service

router = APIRouter(tags=["income"])


# --- 月薪與儲蓄目標 -------------------------------------------------------------------


@router.get("/months/{month}/income", response_model=MonthIncomeOut, responses=error_responses(400))
def get_month_income(month: str, db: DbSession, user_id: CurrentUserId) -> MonthIncomeOut:
    return MonthIncomeOut.model_validate(service.get_month_income(db, user_id, month))


@router.put("/months/{month}/income", response_model=MonthIncomeOut, responses=error_responses(400))
def put_month_income(
    month: str, payload: MonthIncomeUpdate, db: DbSession, user_id: CurrentUserId
) -> MonthIncomeOut:
    return MonthIncomeOut.model_validate(service.upsert_month_income(db, user_id, month, payload))


# --- 額外收入 ----------------------------------------------------------------------------


@router.get("/extra-incomes", response_model=list[ExtraIncomeOut], responses=error_responses(400))
def list_extra_incomes(
    db: DbSession, user_id: CurrentUserId, month: str | None = None
) -> list[ExtraIncomeOut]:
    return [ExtraIncomeOut.model_validate(r) for r in service.list_extra_incomes(db, user_id, month)]


@router.post(
    "/extra-incomes",
    response_model=ExtraIncomeOut,
    status_code=status.HTTP_201_CREATED,
    responses=error_responses(400),
)
def create_extra_income(payload: ExtraIncomeCreate, db: DbSession, user_id: CurrentUserId) -> ExtraIncomeOut:
    return ExtraIncomeOut.model_validate(service.create_extra_income(db, user_id, payload))


@router.put("/extra-incomes/{row_id}", response_model=ExtraIncomeOut, responses=error_responses(400, 404))
def update_extra_income(
    row_id: int, payload: ExtraIncomeUpdate, db: DbSession, user_id: CurrentUserId
) -> ExtraIncomeOut:
    return ExtraIncomeOut.model_validate(service.update_extra_income(db, user_id, row_id, payload))


@router.delete("/extra-incomes/{row_id}", status_code=status.HTTP_204_NO_CONTENT, responses=error_responses(404))
def delete_extra_income(row_id: int, db: DbSession, user_id: CurrentUserId) -> Response:
    service.delete_extra_income(db, user_id, row_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


# --- 固定支出 ----------------------------------------------------------------------------


@router.get("/recurring-expenses", response_model=list[RecurringExpenseOut], responses=error_responses(400))
def list_recurring_expenses(
    db: DbSession, user_id: CurrentUserId, month: str | None = None
) -> list[RecurringExpenseOut]:
    return [RecurringExpenseOut.model_validate(r) for r in service.list_recurring_expenses(db, user_id, month)]


@router.post(
    "/recurring-expenses",
    response_model=RecurringExpenseOut,
    status_code=status.HTTP_201_CREATED,
    responses=error_responses(400),
)
def create_recurring_expense(
    payload: RecurringExpenseCreate, db: DbSession, user_id: CurrentUserId
) -> RecurringExpenseOut:
    return RecurringExpenseOut.model_validate(service.create_recurring_expense(db, user_id, payload))


@router.put(
    "/recurring-expenses/{row_id}", response_model=RecurringExpenseOut, responses=error_responses(400, 404)
)
def update_recurring_expense(
    row_id: int, payload: RecurringExpenseUpdate, db: DbSession, user_id: CurrentUserId
) -> RecurringExpenseOut:
    return RecurringExpenseOut.model_validate(service.update_recurring_expense(db, user_id, row_id, payload))


@router.delete(
    "/recurring-expenses/{row_id}", status_code=status.HTTP_204_NO_CONTENT, responses=error_responses(404)
)
def delete_recurring_expense(row_id: int, db: DbSession, user_id: CurrentUserId) -> Response:
    service.delete_recurring_expense(db, user_id, row_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/income.ts
/** 收入設定 API（SRS 4.1；spec-gaps 8.3、8.4）。型別全部來自 schema.d.ts，不手寫。 */
import { apiFetch, resolvePath, withQuery } from "./client";
import type { components } from "./schema";

export type MonthIncome = components["schemas"]["MonthIncomeOut"];
export type MonthIncomeUpdate = components["schemas"]["MonthIncomeUpdate"];
export type ExtraIncome = components["schemas"]["ExtraIncomeOut"];
export type ExtraIncomeCreate = components["schemas"]["ExtraIncomeCreate"];
export type ExtraIncomeUpdate = components["schemas"]["ExtraIncomeUpdate"];
export type RecurringExpense = components["schemas"]["RecurringExpenseOut"];
export type RecurringExpenseCreate = components["schemas"]["RecurringExpenseCreate"];
export type RecurringExpenseUpdate = components["schemas"]["RecurringExpenseUpdate"];

// --- 月薪與儲蓄目標 ---------------------------------------------------------------------

/** 該月無紀錄時後端回上一有紀錄月份的值並填 inherited_from；完全沒有紀錄時 salary／savings_target 為 null。 */
export function getMonthIncome(month: string): Promise<MonthIncome> {
  return apiFetch<MonthIncome>(resolvePath("/api/v1/months/{month}/income", { month }));
}

export function putMonthIncome(month: string, payload: MonthIncomeUpdate): Promise<MonthIncome> {
  return apiFetch<MonthIncome>(resolvePath("/api/v1/months/{month}/income", { month }), { method: "PUT", json: payload });
}

// --- 額外收入 ----------------------------------------------------------------------------

export function listExtraIncomes(month?: string): Promise<ExtraIncome[]> {
  return apiFetch<ExtraIncome[]>(withQuery("/api/v1/extra-incomes", { month }));
}

export function createExtraIncome(payload: ExtraIncomeCreate): Promise<ExtraIncome> {
  return apiFetch<ExtraIncome>("/api/v1/extra-incomes", { method: "POST", json: payload });
}

export function updateExtraIncome(id: number, payload: ExtraIncomeUpdate): Promise<ExtraIncome> {
  return apiFetch<ExtraIncome>(resolvePath("/api/v1/extra-incomes/{row_id}", { row_id: id }), { method: "PUT", json: payload });
}

export function deleteExtraIncome(id: number): Promise<void> {
  return apiFetch<void>(resolvePath("/api/v1/extra-incomes/{row_id}", { row_id: id }), { method: "DELETE" });
}

// --- 固定支出 ----------------------------------------------------------------------------

/** 不帶 month：全部（設定頁要列出全部，結束月已過的灰階呈現但不隱藏）。 */
export function listRecurringExpenses(month?: string): Promise<RecurringExpense[]> {
  return apiFetch<RecurringExpense[]>(withQuery("/api/v1/recurring-expenses", { month }));
}

export function createRecurringExpense(payload: RecurringExpenseCreate): Promise<RecurringExpense> {
  return apiFetch<RecurringExpense>("/api/v1/recurring-expenses", { method: "POST", json: payload });
}

export function updateRecurringExpense(id: number, payload: RecurringExpenseUpdate): Promise<RecurringExpense> {
  return apiFetch<RecurringExpense>(resolvePath("/api/v1/recurring-expenses/{row_id}", { row_id: id }), { method: "PUT", json: payload });
}

export function deleteRecurringExpense(id: number): Promise<void> {
  return apiFetch<void>(resolvePath("/api/v1/recurring-expenses/{row_id}", { row_id: id }), { method: "DELETE" });
}
